bin/sync_manifest.py

```diff
@@ -0,0 +1,191 @@
+        n = sub_dir.count(os.path.sep) + 1 if sub_dir != '' else 0
+        dirpath = os.path.abspath(dirpath)
+        if depth is not None:
+            if n < depth:
+                dir_list.append(dirpath)
+        else:
+            dir_list.append(dirpath)
+        for name in filenames:
+            file_list.append(os.path.join(dirpath,name))
+    return dir_list,file_list
+
+# walk through Synapse
+# update folders in Synapse to match the local dir
+# get key-value pairs of dirname and synapse id
+def _getSynapseDir(syn, synapse_id, local_root, dir_list):
+    synapse_dir = {}
+
+    synapse_root = syn.get(synapse_id)
+
+    for (dirpath, dirpath_id), _, _ in synu.walk(syn,synapse_id):
+        dirpath = dirpath.replace(synapse_root.name, os.path.abspath(local_root))
+        synapse_dir[dirpath] = dirpath_id
+
+    for di in dir_list:
+        if not synapse_dir.has_key(di):
+            new_folder = Folder(os.path.basename(di),synapse_dir[os.path.dirname(di)])		
+            new_folder = syn.store(new_folder)
+            synapse_dir[di] = new_folder.id
+    return synapse_dir
+
+# get a list of annotation keys
+def _getAnnotationKey(dirs):
+    key_list = ['used','executed']
+    if dirs is not None:
+        for di in dirs:
+            if urlparse.urlparse(di).scheme != '':
+                jfile = urllib.urlopen(di)
+            else:
+                jfile = open(di,'r')
+            base,ext=os.path.splitext(os.path.basename(di))
+            if ext=='.json':
+                data=json.load(jfile)
+            else:
+                sys.stdout.write('File %s cannot be parsed. JSON format is required. \n' % di)
+            data = pd.DataFrame(data)
+            annotation_key = data['name']
+            key_list = key_list + list(annotation_key)
+    return key_list
+
+def _getName(path, synapse_dir, local_root, depth):
+    path_no_root = path[len(os.path.abspath(local_root)):]
+
+    if depth is not None and path_no_root.count(os.path.sep) > depth-1:
+        if str.startswith(path_no_root, '/'):
+            path_no_root = path_no_root[1:]
+        temp_name = path_no_root.split('/')[(depth-1):]
+        name = '_'.join(temp_name)
+    
+        temp_name = '/'.join(temp_name)
+        parent = synapse_dir[os.path.dirname(path[:-len(temp_name)])]
+    else:
+        name = os.path.basename(path)
+        parent= synapse_dir[os.path.dirname(path)]
+    return name,parent
+
+# create manifest
+def create(file_list,key_list,synapse_dir,local_root,depth):
+    result = pd.DataFrame()
+    result['path'] = file_list
+    result['name'] = ""
+    result['parent'] = ""
+    
+    for index, row in result.iterrows():
+        row[['name','parent']] = _getName(row['path'], synapse_dir, local_root, depth)
+        
+    cols = list(result.columns)
+
+    result = pd.concat([result,pd.DataFrame(columns=key_list)])
+    result = result[cols + key_list] # reorder the columns
+
+    result.to_csv('output.tsv',sep = '\t', index=False)
```

Comment:
It's command line standard operating procedure. I would rather see that you have a script that creates necessary Synapse structure and manifest file, and then let the (soon to be implemented) Synapse sync command line take over (`sync_manifest -c -d mydir/ -i syn123 | synapse sync`). Then it's modular, doesn't rely on changes to Synapse sync command, etc.